test/SpagoSpec.hs

```diff
@@ -5,831 +5,880 @@ import qualified Data.Text          as Text
 import           Prelude            hiding (FilePath)
 import qualified System.IO.Temp     as Temp
 import           Test.Hspec         (Spec, around_, describe, it, shouldBe, shouldNotSatisfy,
-                                     shouldNotBe, shouldReturn, shouldSatisfy)
+                                     shouldNotBe, shouldReturn, shouldSatisfy, runIO)
 import           Turtle             (ExitCode (..), cd, cp, decodeString, empty, encodeString,
                                      mkdir, mktree, mv, pwd, readTextFile, rm, shell,
-                                     shellStrictWithErr, testdir, writeTextFile, (</>))
+                                     shellStrictWithErr, testdir, writeTextFile, (</>), die, when, void)
 import           Utils              (checkFileHasInfix, checkFixture, checkFileExist, outputShouldEqual,
                                      readFixture, runFor, shouldBeFailure, shouldBeFailureInfix,
                                      shouldBeFailureStderr, shouldBeSuccess, shouldBeSuccessOutput,
                                      shouldBeSuccessOutputWithErr, shouldBeSuccessStderr, spago,
-                                     withCwd, withEnvVar)
+                                     withCwd, withEnvVar, dhall)
+import qualified Data.Versions as Version
+import qualified Spago.Cmd as Cmd
 
 
-setup :: IO () -> IO ()
-setup cmd = do
-  Temp.withTempDirectory "test/" "spago-test" $ \temp -> do
+setup :: String -> IO () -> IO ()
+setup dir cmd = do
+  Temp.withTempDirectory "test/" dir $ \temp -> do
     -- print ("Running in " <> temp)
     withCwd (decodeString temp) cmd
 
 spec :: Spec
-spec = around_ setup $ do
+spec = do
+  usingEsModules <- runIO $ do
+    pursVersion <- Cmd.getCmdVersion "purs" >>= either die pure
+    esmVersion <- either (const $ die "Failed to parse `0.15.0` static version") pure $ Version.semver "0.15.0-alpha-02"
+    pure $ pursVersion >= esmVersion
+  let
+    purs0_15_0TestMsg = "purs-0.15"
+    whenUsingEsModulesFixPackagesDhall = when usingEsModules $ do
+      -- The prepare-0.15 package set's contents can change.
+      -- So, we copy an unfrozen one into the directory
+      -- and then freeze it before running `spago init`.
+      cp "../fixtures/packages-prepare-0-15.dhall" "packages.dhall"
+      dhall ["freeze", "packages.dhall"] >>= shouldBeSuccess
+
+  around_ (setup "spago-test") $ do
```

Comment:
Adding these few lines here in the beginning is causing a huuge diff, which makes it really hard to review.
Can we do something (e.g. extract this in a function) to try keeping the same indentation as before? We can eventually fix this indentation in a subsequent change.

Reply:
If you look at the commit history, this diff isn't bad.
- https://github.com/purescript/spago/pull/866/commits/b014d28ec4691df6136e37636d7d89599e521e17
- https://github.com/purescript/spago/pull/866/commits/ea010a5a010ebc4067d6409419c3db6f7e28297b
- https://github.com/purescript/spago/pull/866/commits/918ba6b2918069d10161be7ad6d66423467b04f5
- https://github.com/purescript/spago/pull/866/commits/79766363289149bad13e3f0626df111e4f09225e

If you're looking at the entire diff, then yes, I agree. However, I'm not entirely sure how this should be fixed. Could you provide more guidance?

Reply:
Something like this: (disclaimer: I haven't tried compiling it)
```haskell
-- move these to the top level
purs0_15_0TestMsg = "purs-0.15"

fixPackagesDhall usingEsModules = when usingEsModules $ do
    -- The prepare-0.15 package set's contents can change.
    -- So, we copy an unfrozen one into the directory
    -- and then freeze it before running `spago init`.
    cp "../fixtures/packages-prepare-0-15.dhall" "packages.dhall"
    dhall ["freeze", "packages.dhall"] >>= shouldBeSuccess

getUsingEsModules = runIO $ do
    pursVersion <- Cmd.getCmdVersion "purs" >>= either die pure
    esmVersion <- either (const $ die "Failed to parse `0.15.0` static version") pure $ Version.semver "0.15.0-alpha-02"
    pure $ pursVersion >= esmVersion

-- then do a bind-desugar, so that we can keep the `around` on the same line
-- and we don't have to indent everything else:
spec = getUsingEsModules >>= \usingEsModules -> around_ (setup "spago-test") $ do
```

Reply:
That almost works, but then you run into the test setup that needs the space in `spago-test`: https://github.com/purescript/spago/pull/866/files#diff-1be330c42c3288a92800364a45bc6722483cb2700996458be112b464f0fde5d2R775

So, you get code like:
```
spec = getUsingEsModules >>= \usingEsModules -> around_ (setup "spago-test") $ do
  describe "..."
  describe "..."
  
  around_ (setup "spago test") $ do
    describe "..."
```
which is why I indented it.

If you break it into two parts, then this is actually doable

```
spec = do
  usingEsModules <- getUsingEsModules
  noSpacesInParentDir usingEsModules
  spacesInParentDir usingEsModules

noSpacesInParentDir usingEsModules = around_ (setup "spago-test") $ do
  describe "..."
  describe "..."
  
spacesInParentDir usingEsModules =  around_ (setup "spago test") $ do
    describe "..."
```

Since your concern is the diff, I'm going to try that out.

Reply:
It might be worth trying a different approach (though I might be missing something): since all we care about is having a space in the path, why not just create a folder (with a space) and do the thing in there for the few tests that we care about, instead of parameterizing the whole setup?

Reply:
Ok. The diff is definitely much cleaner now.

Reply:
> It might be worth trying a different approach (though I might be missing something): since all we care about is having a space in the path, why not just create a folder (with a space) and do the thing in there for the few tests that we care about, instead of parameterizing the whole setup?

Because your `checkFixture` function assumes that the `fixtures` directory is always reachable via a `../fixtures` path. I suppose you could do a recursive ascent that always checks for a `fixture` directory at the current level and, if it's not there, to go up one level before calling itself again.

Reply:
Ok. That's about as clean of a diff as I can get.